### PR TITLE
Stop type checking in numerical_grad

### PR DESCRIPTION
**Pull request by okuta**

### CODE REVIEW

**Comment by delta2323** [comment]
I would like to know the benefit of this change. Could you tell me the reason why we should and can safely remove type checking?

File: chainer/gradient_check.py [hunk 1]
```diff
@@ -48,18 +44,22 @@ def numerical_grad(f, inputs, grad_outputs, eps=1e-3):
...
             x[i] = orig - eps
-            ys2 = _copy_arrays(f())
+            ys2 = [xp.copy(j) for j in f()]
             x[i] = orig
-            for y1, y2, gy in zip(ys1, ys2, grad_outputs):
+            for y1, y2, gy in six.moves.zip(ys1, ys2, grad_outputs):
                 if gy is not None:
                     dot = ((y1 - y2) * gy).sum()
                     gx[i] += dot / (2 * eps)
+    function.Function.type_check_enable = prev_mode
```
Comment (gwtnb) [inline chainer/gradient_check.py]: If an exception is raised before this statement, `function.Function.type_check_enable` is not restored.

**Comment by okuta** [comment]
This PR speeds up tests.

Before
｀｀｀
% nosetests tests/chainer_tests/functions_tests -A "gpu<2"
Ran 6354 tests in 123.719s
｀｀｀

After
｀｀｀
% nosetests tests/chainer_tests/functions_tests -A "gpu<2"
Ran 6354 tests in 89.673s
｀｀｀

File: chainer/gradient_check.py (hunk 1 above)
Comment (okuta) [inline chainer/gradient_check.py]: I use `finally` to fix.

**Comment by gwtnb** [comment]
```
ERROR: test_differenct_eps_gpu (test_gradient_check.NumericalGradientEpsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/chainer/tests/chainer_tests/test_gradient_check.py", line 255, in test_differenct_eps_gpu
    self.check_different_eps(cuda.to_gpu(self.x), cuda.to_gpu(self.y))
  File "/tmp/chainer/tests/chainer_tests/test_gradient_check.py", line 247, in check_different_eps
    gx, = gradient_check.numerical_grad(f, (x,), (y,), eps=2.5)
  File "/tmp/chainer/chainer/gradient_check.py", line 55, in numerical_grad
    ys1 = [xp.copy(j) for j in f()]
  File "/tmp/chainer/chainer/gradient_check.py", line 55, in <listcomp>
    ys1 = [xp.copy(j) for j in f()]
  File "/tmp/chainer/cupy/creation/from_data.py", line 104, in copy
    return a.copy()
AttributeError: 'int' object has no attribute 'copy'

======================================================================
FAIL: test_numerical_grad_gpu (test_gradient_check.NumericalGradientTest2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/chainer/chainer/testing/condition.py", line 71, in wrapper
    fail()
  File "/tmp/chainer/chainer/testing/condition.py", line 51, in fail
    instance.fail(msg)
AssertionError:
Fail: 3, Success: 0

The first error message:
Traceback (most recent call last):
  File "/tmp/chainer/chainer/testing/condition.py", line 57, in <lambda>
    lambda: f(*args, **kwargs),
  File "/tmp/chainer/tests/chainer_tests/test_gradient_check.py", line 84, in test_numerical_grad_gpu
    eps=self.eps)
  File "/tmp/chainer/tests/chainer_tests/test_gradient_check.py", line 69, in check_numerical_grad
    self.check_numerical_grad_one(f, df, xs, gys, e)
  File "/tmp/chainer/tests/chainer_tests/test_gradient_check.py", line 55, in check_numerical_grad_one
    dx_actual = gradient_check.numerical_grad(func, xs, gys, eps)
  File "/tmp/chainer/chainer/gradient_check.py", line 55, in numerical_grad
    ys1 = [xp.copy(j) for j in f()]
  File "/tmp/chainer/chainer/gradient_check.py", line 55, in <listcomp>
    ys1 = [xp.copy(j) for j in f()]
  File "/tmp/chainer/cupy/creation/from_data.py", line 104, in copy
    return a.copy()
AttributeError: 'int' object has no attribute 'copy'
```

**Comment by unnonouno** [comment]
Is it safe in multi-threading environment?

**Comment by okuta** [comment]
This functionality is not thread safe.
But, current test runs on single thread or multi process.

**Comment by unnonouno** [comment]
Please write a comment about test environment. I mean, "Test scripts always run in single thread or multi-process."

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**